### ISSUE TITLE
conflict between `use_key_in_widget_constructors`, `unused_element `, and `use_super_parameters`

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/linter#3314]
The following code produces a `unused_element` warning for the `key` parameter (note that the class is private):

```dart
class _MyWidget extends StatelessWidget {
  const _MyWidget({super.key, required this.child});

  final Widget child;
  
  @override
  Widget build(BuildContext context) => child;
}
```

Removing the `key` parameter produces a `use_key_in_widget_constructors` warning:

```dart
class _MyWidget extends StatelessWidget {
  const _MyWidget({required this.child});

  final Widget child;
  
  @override
  Widget build(BuildContext context) => child;
}
```

There is no warning if you don't use the new super syntax for the `key` param (this will however trigger the new `use_super_parameters` lint):

```dart
class _MyWidget extends StatelessWidget {
  const _MyWidget({Key? key, required this.child}) : super(key: key);

  final Widget child;
  
  @override
  Widget build(BuildContext context) => child;
}
```

**Comment by goderbauer** [comment]
Either `use_key_in_widget_constructors` shouldn't be applied to private widgets or `unused_element` shouldn't apply to the key parameter in this case.

I think we would want all Widgets to have a `key` parameters (private or not) for consistency and ease of use (just like the world was prior to introducing the super parameter feature).

/cc @Hixie @mit-mit @pq @srawlins

**Comment by pq** [comment]
> Either use_key_in_widget_constructors shouldn't be applied to private widgets 

I believe this is currently the case.

For example,

```dart
class _MyWidget extends StatelessWidget {
  final Widget child;
  const _MyWidget({required this.child});

  @override
  Widget build(BuildContext context) => Container();
}
```

does not produce a lint for me but

```dart
class MyWidget extends StatelessWidget {
  final Widget child;
  const MyWidget({required this.child});

  @override
  Widget build(BuildContext context) => Container();
}
```

does.

Could you double check what you're seeing on your end?

**Comment by pq** [comment]
`use_key_in_widget_constructors` should also skip private constructors in public classes.  For example

```dart
class MyWidget extends StatelessWidget {
  final Widget child;
  const MyWidget._({required this.child});

  @override
  Widget build(BuildContext context) => Container();
}
```

analyzes without a lint for me.

**Comment by goderbauer** [comment]
You are right. `use_key_in_widget_constructors` doesn't actually apply to private classes. Looks like this was a false alarm. Not sure what I was seeing earlier; when I try it again just now this analyzes without error:

```dart
class _MyWidget extends StatelessWidget {
  const _MyWidget({required this.child});

  final Widget child;
  
  @override
  Widget build(BuildContext context) => child;
}
```

I'll close this.
